test/e2e/scheduling/predicates.go

```diff
@@ -356,7 +356,7 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
 		framework.ExpectNoError(err)
 		for _, pod := range pods.Items {
 			_, found := nodeToAllocatableMap[pod.Spec.NodeName]
-			if found && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
+			if found && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed && pod.DeletionTimestamp == nil {
```

Comment:
This might cause another flake where the last pod is scheduled if one Pod with DeletionTimestamp doesn't ever terminate.

Have you tried increasing timeouts in the cleanup function for replication controller?

Reply:
> This might cause another flake where the last pod is scheduled if one Pod with DeletionTimestamp doesn't ever terminate.

If the pod with DeletionTimestamp set does not terminate, there will be even more resource requested. I.e. resource requested by all pods without DeletionTimestamp set + filler pods + pods with DeletionTimestamp set.

Reply:
Sorry, the case I was thinking of is actually this:

There are a lot of Pods with DeletionTimestamp initially, and they they all terminate. Thus there is a lot of space and the last pod gets scheduled.

Reply:
Let's use `pods` for a set of all pods (including pods with `DeletionTimestamp` set) and `podswithoutDT` for a set of all pods with `DeletionTimestamp` set to nil

```
for _, pod := range pods.Items {
	_, found := nodeToAllocatableMap[pod.Spec.NodeName]
	if found && pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed && pod.DeletionTimestamp == nil {
		framework.Logf("Pod %v requesting resource cpu=%vm on Node %v", pod.Name, getRequestedCPU(pod), pod.Spec.NodeName)
		nodeToAllocatableMap[pod.Spec.NodeName] -= getRequestedCPU(pod)
	}
}
```

After the cycle `nodeToAllocatableMap[node]` constructed from `pods` will be less or equal to the same map constructed from `podswithoutDT`. I.e. `nodeToAllocatableMap_pods[node] <= nodeToAllocatableMap_podswithoutDT[node]`.

With the next cycle where the `filler-pod-XXX` pods are constructed:
```
for nodeName, cpu := range nodeToAllocatableMap {
	requestedCPU := cpu * 7 / 10
...
```

Given `nodeToAllocatableMap_pods[node] <= nodeToAllocatableMap_podswithoutDT[node]`, also `requestedCPU` for set `pods` will be less or equal to `podswithoutDT`. I.e.

`requestedCPU_pods <= requestedCPU_podswithoutDT`.

In other words, requested cpu of a filler pod for all pods including pods with DeletionTimestamp set (on a given node) will never be more than requested cpu for all pods (on a given node) without pods with DeletionTimestamp set. So if any of the pods with DeletionTimestamp set disappears, the requested cpu (from `podswithoutDT` set) of a filler pod will compensate. 

It's actually the opposite problem in question. What if the pods with DeletionTimestamp set do not disappear before the `additional-pod` is scheduled? `additional-pod` will not still get scheduled since there's gonna be even less of cpu resource left. The worst case scenario is when there's so many pods with DeletionTimestamp set that even the filler pods will not get scheduled.

Reply:
> There are a lot of Pods with DeletionTimestamp initially and they all terminate after this loop. Thus there is a lot of space and the last pod gets scheduled.

This is actually the case happening now

Reply:
Thanks for detailed analysis. You are right, the scenario where `additional-pod` doesn't get scheduled is possible. It sounds like the only correct solution is to fix the cleanup of other tests :/

Reply:
I am afraid there can be other tests not cleaning their pods properly in the future (or maybe even now). If so, this needs to be fixed on multiple places. I'd start here since this change is making the test less flaky if there's not so many pods with DeletionTimestamp set.

Reply:
I'm not comfortable with this patch. @Huang-Wei @ahg-g to comment.

Reply:
Even if this makes the test less flaky, that's not necessarily better, as we might be hiding legitimate bugs.

Reply:
>  The worst case scenario is when there's so many pods with DeletionTimestamp set that even the filler pods will not get scheduled.

Right, and this will most likely cause flakes in the future.

We already call `WaitForStableCluster`, perhaps we can modify it to call `WaitForPodsToBeDeleted`? If those pods don't get deleted, then there is something wrong happening and we can address that instead of working around it by ignoring pods to be deleted.

Reply:
> We already call `WaitForStableCluster`, perhaps we can modify it to call `WaitForPodsToBeDeleted`?

+1, instead of embedding `WaitForPodsToBeDeleted` into `WaitForStableCluster`, another option is to call `WaitForPodsToBeDeleted` on demand, immediately after `WaitForStableCluster`, for example in the case.

BTW: `WaitForPodsToBeDeleted` was used in an old test case (which is deleted now) in 1.17:

https://github.com/kubernetes/kubernetes/blob/5ec472285121eb6c451e515bc0a7201413872fa3/test/e2e/scheduling/predicates.go#L132

So I _guess_ `WaitForPodsToBeDeleted` was introduced at that time to resolve similar flakes.

Reply:
Another thing a bit off the topic: both `WaitForStableCluster` and `WaitForPodsToBeDeleted` are performing an infinite `for` loop, although the ginkgo test may be guarded by a timeout argument, I think it'd be good to explicitly time out the test inside the function. Anyway, it can be a follow-up.

Reply:
I was able to reproduce the test locally. The remaining pods (`67e791a6-f170-4aff-b2b1-e42a0a6a7c3c-0`, `ead2872e-4ad3-4244-bb02-9f7ff0b6a428-0` and `ffecd964-0982-4541-ba35-bc53ef1bcd52-0`) were not deleted before the testing namespace was. Thus, leaving the pods still running even after the test ended. PR update to delete all remaining pods in the testing namespace first.

@alculquicondor @ahg-g @Huang-Wei  PTAL

Reply:
I still have the same issue with this patch.

Why don't we fix `DeleteRCAndWaitForGC` instead?

Reply:
The remaining pods have no owner

Reply:
can you tell what created them then?

Again, I'd like to fix the root of the problem.

Reply:
https://github.com/kubernetes/kubernetes/blob/64c099d67069f85f2b9b78d7d12fd1a92fcc3d75/test/e2e/scheduling/priorities.go#L484

Other option is to have `createBalancedPodForNodes` return cleanup function that will be called in `defer` block.

Reply:
oh, I get it now. These Pods are not created for the RC.

2 things then:
- Use `DeleteCollection()`.
- Do this in a separate defer, right after the Pods are created. Otherwise if RC creation fails, these Pods don't get cleaned.

Maybe for follow up: could you check for other instances where `createBalancedPodForNodes` is called and Pods are not cleared? Is this ever intended?

Reply:
`DeleteCollection` does not wait until all the pods are deleted. Which makes it unusable without additional waiting logic.

Reply:
PR updated. `createBalancedPodForNodes` now returns cleanUp function which is called right away through defer.